### PR TITLE
Immutability: analyzer assumes initializers are the only writes to readonly members #319

### PR DESCRIPTION
**Pull request by TimothyJCowen**
- Added method to get all possible initializations from within constructors
- Changed single-initializer structures to use multiple initializers
- Added supporting tests

### CODE REVIEW

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 1]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
 				diagnosticSink( diagnostic );
 				immutable = false;
 			}
 
 			return immutable;
 		}
 
+
+		/// <summary>
+		/// A readonly field or property may be re-initialized in constructors.
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:267]: Nit: I get what you mean here by initializers but it's a bit confusing because initializers are a specific thing in the language. Maybe reword this to writes? Like `GetWrites()` etc. which sounds more general/accurate.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 2]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+					initializer
+				);
 
-			if ( stuff == null ) {
-				// null is a signal that there is nothing further that needs to
-				// be checked
-				return immutable;
-			}
+				if( stuff == null ) {
+					// Null is a signal that there is nothing further that needs to
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:242]:
```suggestion
					// null is a signal that there is nothing further that needs to
```

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 3]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+		/// <summary>
+		/// A readonly field or property may be re-initialized in constructors.
+		/// This method finds all initializations of the field/property.
+		/// </summary>
+		/// <param name="symbol">The field/property to search for</param>
+		/// <param name="initialInitializer">The initialization with the declaration</param>
+		/// <returns>An immutable list of initializations</returns>
+		private ImmutableArray<ExpressionSyntax> GetInitializers(
+			ISymbol symbol,
+			ExpressionSyntax initialInitializer
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:275]: heh ya and rename this to initializer :)

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 4]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+		) {
+			List<ExpressionSyntax> initializers = new() {initialInitializer};
+
+			// If the containing symbol isn't a named type (class) then there
+			// will be no constructors
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:280]: Members can exist in other INamedTypeSymbols, like records or structs.
What is an example of a field that is not an INamedTypeSymbol?

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 5]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+			if( symbol.ContainingSymbol is not INamedTypeSymbol namedTypeSymbol ) {
+				return initializers.ToImmutableArray();
+			}
+
+			// Iterate through all constructors in the class
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:285]:
```suggestion
```

Yes, its a for loop over the constructors, the code already says that :)

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:285]: https://blog.d2l.dev/2019/08/29/write-good-comments/

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 6]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+			foreach( IMethodSymbol methodSymbol in namedTypeSymbol.Constructors ) {
+				// Make sure the syntax of the constructor exists
+				if( methodSymbol.DeclaringSyntaxReferences.Length == 0 ) {
+					continue;
+				}
+
+				// Retrieve the constructor syntax
+				var methodSyntax = methodSymbol.DeclaringSyntaxReferences
+				                               .Single()
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: We don't align `.`s like this iirc

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: If you put `.DeclaringSyntaxReferences on the next line then they will line up, that's pretty common.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 7]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+				                               .GetSyntax() as ConstructorDeclarationSyntax;
+
+				// Make sure the constructor actually contains a body
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:297]: Are there constructors without a body?

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 8]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+				if( methodSyntax?.Body?.Statements == null ) {
+					continue;
+				}
+
+				// Retrieve the semantic model from the compilation
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:302]:
```suggestion
```

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 4 above)
Comment (TimothyJCowen) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:280]: The check is because it is `INamedTypeSymbol` that has the `.Constructors` member.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 9]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+				SemanticModel semanticModel = m_compilation.GetSemanticModel( methodSyntax.SyntaxTree );
+
+				// Iterate through every statement in the constructor
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:305]:
```suggestion
```

We don't need comments that just rephrase a line of code as english.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 6 above)
Comment (TimothyJCowen) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: The IDE does. I've never seen it done a different way.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 10]
```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
...
+				foreach( StatementSyntax statement in methodSyntax.Body.Statements ) {
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:306]:
```suggestion
				foreach( var statement in methodSyntax.Body.Statements ) {
```

Can you go though the rest of the change and do this one? In the analyzer code we use `var` by default.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 7 above)
Comment (TimothyJCowen) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:297]: You wouldn't expect so, but the IDE was warning me of potential `NullReferenceException`s so I figured it's best to listen to it and be safe.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 10 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:306]: Actually, this doesn't work:

```cs
public MyType() {
  if ( condition ) // x
  { // y
    m_foo = bar; // z
  }
}
```

The statement `z` isn't in `methodSyntax.Body.Statements`, it's a child statement of the block `y` which is attached to the (if) statement `x`.

There is an easier way to do this than nested loops though (you'd need recursion/a stack to fix this up):

```cs
var writes = namedTypeSymbol
    .Constructors
    .SelectMany( c => c.DescendentNodes() )
    .OfType<AssignmentExpressionSyntax>();
```

It's not as efficient as it could be, because there are cases where we would never find a descendent `AssignmentExpressionSyntax` (like inside of a `case` pattern, like you can't do `case (m_foo = bar):`) but it would be hard to actually implement that and possible fragile (the language can change).

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 6 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:291]: Resharper? Mine doesn't 🤔

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 7 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: Hm, I wouldn't listen to the IDE then.

The problem with quietly backing away like this in the analyzer when we find unexpected syntax is it creates room for bugs. The code here is saying "well, if it doesn't have a body I guess stuff is fine.". What if there is a form of constructor that doesn't have a body but does write to fields? It doesn't seem likely to me, but I would rather we crashed with a NullReferenceException :)

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: In cases where we are less sure we have made diagnostics, e.g.: https://github.com/Brightspace/D2L.CodeStyle/blob/master/src/D2L.CodeStyle.Analyzers/Diagnostics.cs#L473-L481

This one makes sense because C# may add a new type kind in the future.

For this case, it seems unlikely to me that there is going to be new syntax for bodyless constructors (I could be wrong, maybe it exists now!) so I'm happy with a NullReferenceException (its simpler.)

Comment (omsmith) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: I believe it's `Body` vs `ExpressionBody` (arrow syntax). They're mutually exclusive and the other is null.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 11]
```diff
@@ -249,33 +339,29 @@ ExpressionSyntax initializer
...
 		/// <param name="memberType">The type of the field/property</param>
 		/// <param name="memberTypeSyntax">The syntax for the type of the field/property</param>
-		/// <param name="initializer">The initializer syntax for the field/property (possibly null)</param>
+		/// <param name="assignment">The assignment syntax for the field/property (possibly null)</param>
 		/// <returns>null if no checks are needed, otherwise a bunch of stuff.</returns>
 		private (ITypeSymbol, ImmutableTypeKind, Func<Location>)? GetStuffToCheckForMember(
 			ITypeSymbol memberType,
 			TypeSyntax memberTypeSyntax,
-			ExpressionSyntax initializer
+			ExpressionSyntax assignment
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:347]: This one actually is the initializer :)

Looking at this method in isolation, if I were to call it, I'd be wondering "which assignment"?
When it's called `initializer` I know exactly what it means and how to call it.

Comment (TimothyJCowen) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:347]: It's actually all of them. This method gets called for each assignment, not just the initializer. See `CheckFieldOrProperty(...)`.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 7 above)
Comment (omsmith) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: So that is to say, we need to handle `ExpressionBody`.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 4 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:278]: Yeah, so this is similar to the `NullReferenceException`. There are two ways to deal with the cast:

1. Assert that it always exists and suffer an `InvalidCastException` if you're wrong (which will trigger a bug report against the analyzers)
2. Handle the failure to cast somehow

If we do (2) and just return an empty array we are saying that there are no extra writes for non-`INamedTypeSymbol`... so I'm wondering if there is any case that can happen? Does returning `[]` make sense in all those cases? Is there a chance that we miss assignments by doing this, which would be a bug?

If we think `INamedTypeSymbol` is the only option here we should do option (1)

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 7 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: Are expression bodies supported for constructors?

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: Oh damn they are.

Aside: we should have test cases for the destructuring assignment, that might be its own case I'm not sure. Probably bettter as a follow-up PR though.

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: `foo.Body ?? foo.ExpressionBody`

Comment (omsmith) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:294]: > Oh damn they are

For specific example: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members#constructors

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs (hunk 11 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:347]: Ohh yes good stuff.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 12]
```diff
@@ -219,28 +228,109 @@ ExpressionSyntax initializer
...
+					typeToCheck,
+					checkKind,
+					getLocation,
+					out var diagnostic
+				) ) {
+					continue;
+				}
 
-			if( !m_context.IsImmutable( typeToCheck, checkKind, getLocation, out var diagnostic ) ) {
 				diagnosticSink( diagnostic );
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:256]: So we'll output multiple diagnostics here potentially.... I like it 👍🏼

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 13]
```diff
@@ -219,28 +228,109 @@ ExpressionSyntax initializer
...
+		) {
+			List<ExpressionSyntax> assignments = new() {initializer};
+
+			// If the containing symbol isn't a named type (class) then there
+			// will be no constructors
+			if( symbol.ContainingSymbol is not INamedTypeSymbol namedTypeSymbol ) {
+				return assignments.ToImmutableArray();
+			}
+
+			foreach( var methodSymbol in namedTypeSymbol.Constructors ) {
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:283]: It got lost in the shuffle, but: https://github.com/Brightspace/D2L.CodeStyle/pull/698#discussion_r566178258

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 14]
```diff
@@ -219,28 +228,87 @@ ExpressionSyntax initializer
...
+			// If the containing symbol isn't a named type (class) then there
+			// will be no constructors
+			if( symbol.ContainingSymbol is not INamedTypeSymbol namedTypeSymbol ) {
+				return assignments.ToImmutableArray();
+			}
+
+			// Retrieve syntax of all constructors
+			foreach(
+				var constructorSyntax in
+				from syn in
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:286]: Ah, please don't use this syntax, we don't really use it at D2L.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 15]
```diff
@@ -219,28 +228,87 @@ ExpressionSyntax initializer
...
+					from sym in namedTypeSymbol.Constructors
+					where sym.DeclaringSyntaxReferences.Length != 0
+					select sym.DeclaringSyntaxReferences
+					          .Single()
+					          .GetSyntax() as ConstructorDeclarationSyntax
+				where syn != null
+				select syn
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:293]: The DeclaringSyntaxReference for any non-implicit (I think `.IsImplicit` is on the symbol) constructor is always a `ConstructorDeclarationSyntax`, no need for the `where`

Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:293]: http://sourceroslyn.io/#Microsoft.CodeAnalysis/Symbols/ISymbol.cs,5dd9c88c9828d99d,references

You can put this in a `.Where` to filter those out.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 16]
```diff
@@ -274,38 +274,50 @@ ExpressionSyntax initializer
...
+			// Retrieve a list of assignment expressions from any constructors
+			// within the class
+			var assignmentExpressions = namedTypeSymbol
+			                           .Constructors
+			                           .Where( sym => !sym.IsImplicitlyDeclared )
+			                           .Select( sym => sym.DeclaringSyntaxReferences
+			                                              .Single()
+			                                              .GetSyntax() )
+			                           .SelectMany( syn => syn.DescendantNodes() )
+			                           .OfType<AssignmentExpressionSyntax>();
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:303]: Ok, this looks good, a few things:

* nit: the right alignment of the `.` here is wrong. Is this resharper, or did you configure your editor to do this? Your editor should [be doing this](https://github.com/Brightspace/D2L.CodeStyle/blob/master/src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs#L34-L41). It's not a big deal, it's just that if everyone disagrees then every PR is going to constantly fight between styles. Consistency is key.
* It's probably fine to combine the and `Select`+`Where` and `Select` from below on to the end of here, but I would combine the `Select`+`Where` into a single `Select` with a helper, e.g.

```cs
.Where( ae => IsAnAssignmentTo( ae, symbol ) )
.Select( ae => ae.Right );
```

The first `Select` that creates an anonymous type to coordinate stuff between the next `Where` and `Select` makes it hard to read.

* In `IsAnAssignmentTo`, instead of passing the semantic model down, grab it each time based on the passed in assignment expression. What we're doing here I think has a bug, because `partial` classes means that the semantic model for the field/property may not have include the constructor that writes to it (because it could be in a different file/document)
* nit: it's a bit weird to create a single element `List`, compose it into an `IEnumerable` with `AddRange` and then convert it to `ImmutableArray`. You can do `assignments.Append( initializer ).ToImmutableArray()` to achieve the same end result without the extra allocation.

File: src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs [hunk 17]
```diff
@@ -274,38 +274,50 @@ ExpressionSyntax initializer
...
+			var semanticModel = m_compilation.GetSemanticModel(
+				symbol.DeclaringSyntaxReferences.Single().GetSyntax().SyntaxTree
+			);
+
+			// Retrieve a list of assignment expressions from any constructors
+			// within the class
+			var assignmentExpressions = namedTypeSymbol
+			                           .Constructors
+			                           .Where( sym => !sym.IsImplicitlyDeclared )
+			                           .Select( sym => sym.DeclaringSyntaxReferences
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs:299]: one more nit: we have `symbol` in this method that refers to a property, `namedTypeSymbol` which refers to its containing type, and `sym` here which refers to a constructor... we should probably choose better names for readability (member/containingType/constructor (or just c inside these lambdas))

File: tests/D2L.CodeStyle.Analyzers.Test/Specs/ImmutabilityAnalyzer.cs [hunk 18]
```diff
@@ -54,6 +55,40 @@ public interface RegularInterface { }
...
+
+		[Immutable]
+		public sealed class SomeOtherOtherClassWithConstructor {
+			public readonly RegularInterface m_interface = new Good();
+
+			public SomeClassWithConstructor() {
+				m_interface = new Good();
+			}
+		}
+		#endregion
```
Comment (j3parker) [inline tests/D2L.CodeStyle.Analyzers.Test/Specs/ImmutabilityAnalyzer.cs:90]: Other test cases that would be nice:

* one with a more complicated constructor, like one with writes inside an if statement (to catch the previous bug that `DescendantNodes()` fixed
* expression-bodies constructors (`public Mytype() => ...`)
* destructuring assignment

The last two might not work yet, and if so then they can be implemented in a follow-up (this PR has had enough back and forth :) ) I got this example from a Microsoft blog that hits the last two cases:

```cs
public record Person { 
    public string FirstName { get; init; } 
    public string LastName { get; init; }

    public Person(string firstName, string lastName) 
      => (FirstName, LastName) = (firstName, lastName);
}

Comment (j3parker) [reply tests/D2L.CodeStyle.Analyzers.Test/Specs/ImmutabilityAnalyzer.cs:90]: Maybe throw this in as a test case, it should pass now:

```cs
public record Person(string FirstName, string LastName);
```

It's equivalent to the previous, but it's constructor is implicit. We shouldn't analyze it's constructor, but that's fine because the writes inside of it are always safe.

**Review by j3parker — DISMISSED** [review]

**Review by j3parker — APPROVED** [review]
